src/netlink.c

```diff
@@ -0,0 +1,272 @@
+    }
+
+    switch (nl_message->nlmsg_type) {
+        case RTM_DELLINK:
+            enftun_log_info("Network interface %s was removed\n", if_name);
+            nl->on_change(nl);
+            break;
+
+        case RTM_NEWLINK:
+            enftun_log_info("New network interface %s, state: %s %s\n", if_name, if_up_flag, if_run_flag);
+            nl->on_change(nl);
+            break;
+    }
+
+ignore_interface_link:
+    return 0;
+}
+
+int
+enftun_netlink_loop(struct enftun_netlink* nl)
+{
+    (void) nl;
+    ssize_t bytes_in_msg = recvmsg(nl->fd, &nl->msg, MSG_DONTWAIT);
+    if (bytes_in_msg < 0 || (nl->msg.msg_namelen != sizeof(nl->sock_addr))) {
+        return -1;
+    }
+
+    struct nlmsghdr* nl_message;
+    nl_message = (struct nlmsghdr*)nl->buffer;
+
+    while (bytes_in_msg >= (ssize_t)sizeof(*nl_message)) {
+        int length_msghdr = nl_message->nlmsg_len;
+        int length_msg = length_msghdr - sizeof(*nl_message);
+
+        if ((length_msg < 0) || (length_msghdr > bytes_in_msg)) {
+            enftun_log_error("Invalid message length: %i\n", length_msghdr);
+            continue;
+        }
+        int rc = 0;
+        if ((nl_message->nlmsg_type == RTM_NEWROUTE) || (nl_message->nlmsg_type == RTM_DELROUTE)){
+            rc = handle_route_change(nl);
+        } else if ((nl_message->nlmsg_type == RTM_NEWADDR) || (nl_message->nlmsg_type == RTM_DELADDR)) {
+            rc = handle_addr_change(nl, nl_message);
+        } else if ((nl_message->nlmsg_type == RTM_NEWLINK) || (nl_message->nlmsg_type == RTM_DELLINK)) {
+            rc = handle_link_change(nl, nl_message);
+        }
+        if (0 != rc){
+            return -1;
+        }
+
+        // align offsets by the message length, this is important
+        bytes_in_msg -= NLMSG_ALIGN(length_msghdr);
+
+        nl_message = (struct nlmsghdr*)((char*)nl_message + NLMSG_ALIGN(length_msghdr));
+    }
+    return 0;
+}
+
+static
+void
+on_poll(uv_poll_t* handle, int status, int events)
+{
+    (void) status;
+    (void) events;
+    struct enftun_netlink* nl = handle->data;
+
+    if (status < 0)
+        return;
+    if (0 == status)
+        enftun_netlink_loop(nl);
+}
+
+int
+enftun_netlink_init(struct enftun_netlink* nl, uv_loop_t* loop, void* ctx, char* tun_name)
+{
+    nl->fd = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
+    if (nl->fd < 0) {
+        return -1;
+    }
+
+    nl->io_vector.iov_base = nl->buffer;
+    nl->io_vector.iov_len = sizeof(nl->buffer);
+
+    memset(&nl->sock_addr, 0, sizeof(struct sockaddr_nl));
+
+    nl->sock_addr.nl_family = AF_NETLINK;
+    nl->sock_addr.nl_groups = RTMGRP_LINK | RTMGRP_IPV4_IFADDR | RTMGRP_IPV4_ROUTE;
+    nl->sock_addr.nl_pid = getpid();
+
+    nl->msg.msg_name = &nl->sock_addr;
+    nl->msg.msg_namelen = sizeof(nl->sock_addr);
+    nl->msg.msg_iov = &nl->io_vector;
+    nl->msg.msg_iovlen = 1;
+
+    if (bind(nl->fd, (struct sockaddr*)&nl->sock_addr, sizeof(struct sockaddr_nl)) < 0) {
```

Comment:
I'm concerned about a race condition with this not being opened until here.

From `enftun.c`, the current execution sequence looks something like this
1) enftun_connect
    1) open and connect TCP socket (enftun_connect)
    1) open TUN device (enftun_connect)
1) enftun_tunnel
    1) open and bind netlink socket

If the network changes between steps 1.i and 2.i, we'll miss those netlink events.  We need to start listening to netlink events before connecting the TCP socket.

I suggest splitting `enftun_netlink_init` into two functions:

### `enftun_netlink_init`
Does any static initialization, if any.  Does not create or bind the netlink socket.  This function may be empty.

Called from `enftun_context_init`.

`enftun_netlink_free` should undo anything done in this function.

### `enftun_netlink_bind`
Creates and binds the netlink socket.  Can call `uv_poll_init` too.

Called from `enftun_connect` before doing `enftun_tcp_connect`.

`enftun_netlink_close` should undo anything done in this function.

### `enftun_netlink_start`

Keep the same.

`enftun_netlink_stop` undoes everything done in this function.